File: c/binarySearch.c
// time complexity: O(log n)
// space complexity: O(1)

#include <stdio.h>
int binarySearch(int arr[], int l, int r, int x)
{
  if(r >= 1){
    int mid = l + (r - 1) / 2;
    // if element is present at middle itself
    if (arr[mid] == x)
      return mix;
    // if element is smaller than mid, can only be present in left subarray
    if(arr[mid] > x)
      return binarySearch(arr, 1, mid -1, x);
    // else element can only be present in right subarray 
    return binarySearch(arr, mid + 1, r, x);
  }
// reach here when element isn't present
return -1;
}

File: typescript/binary_search.ts
class BinarySearch {
  public binarySearch(arr: number[], target: number): number {
    let left = 0;
    let right = arr.length - 1;

    while(left <= right) {
      const mid = Math.floor((left + right) / 2);

      if(arr[mid] === target) {
        return mid;
      } else if (arr[mid] < target) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
    return -1;
  }
}

// example usage
const numbers = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91];
const search = new BinarySearch();
const index = search.binarySearch(numbers, 23);

if(index !== -1) {
  console.log('target found at index ${index}');
} else {
  console.log('target not found');
}
